pkg/informer/informer.go

```diff
@@ -244,140 +300,103 @@ func (d *DynamicDiscoverySharedInformerFactory) AddEventHandler(handler GVREvent
 	d.handlersLock.Unlock()
 }
 
-func (d *DynamicDiscoverySharedInformerFactory) AddIndexers(indexers cache.Indexers) error {
-	if d.indexers == nil {
-		d.indexers = map[string]cache.IndexFunc{}
-	}
-	for name, indexer := range indexers {
-		if _, found := d.indexers[name]; found {
-			return fmt.Errorf("indexer %q already exists", name)
-		}
-		d.indexers[name] = indexer
-	}
-
-	return nil
-}
+// StartWorker starts the worker that waits for notifications that informer updates are needed. This call is blocking,
+// stopping when ctx.Done() is closed.
+func (d *DynamicDiscoverySharedInformerFactory) StartWorker(ctx context.Context) {
+	defer func() {
+		d.informersLock.Lock()
 
-// StartPolling starts the polling process that periodically discovers new resources and starts informers for them.
-// This call is non-blocking.
-func (d *DynamicDiscoverySharedInformerFactory) StartPolling(ctx context.Context) {
-	// Immediately discover types and start informing.
-	if err := wait.PollImmediateInfiniteWithContext(ctx, time.Second, func(ctx context.Context) (bool, error) {
-		if err := d.discoverTypes(ctx); err != nil {
-			klog.Errorf("Error discovering initial types: %v", err)
-			return false, nil
+		for _, stopCh := range d.informerStops {
+			close(stopCh)
 		}
-		return true, nil
-	}); err != nil {
-		klog.Errorf("Error discovering initial types: %v", err)
-		return
-	}
-
-	// Poll for new types in the background.
-	ticker := time.NewTicker(d.pollInterval)
-	go func() {
-		defer func() {
-			d.mu.Lock()
-			defer d.mu.Unlock()
-
-			// tear down all informers when done.
-			for _, stopCh := range d.informerStops {
-				close(stopCh)
-			}
 
-			// Note: it does not matter if after this another informer is added. It won't be started without
-			// calling discoverTypes.
-		}()
-
-		for {
-			select {
-			case <-ctx.Done():
-				ticker.Stop()
-				return
-			case <-ticker.C:
-				if err := d.discoverTypes(ctx); err != nil {
-					klog.Errorf("Error discovering types: %v", err)
-				}
-			}
-		}
+		d.informersLock.Unlock()
 	}()
-}
 
-func (d *DynamicDiscoverySharedInformerFactory) discoverTypes(ctx context.Context) error {
-	latest := map[schema.GroupVersionResource]struct{}{}
+	wait.UntilWithContext(ctx, func(ctx context.Context) {
```

Comment:
What's the intent with `UntilWithContext` here? Couldn't you just:


```go
for {
    select {
    case <-ctx.Done():
        return
    case <-d.updateCh:
    }
    d.updateInformers()
}
```

Reply:
Gotcha.

Reply:
Added a comment, PTAL